cmake/modules/user_cache.cmake

```diff
@@ -0,0 +1,30 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# Configure user cache directory.
+#
+# The user cache can be used for caching of data that should be persistent
+# across builds to speed up CMake configure / build system generation and/or
+# compilation.
+#
+# Only data that can be safely re-generated should be placed in this cache.
+#
+# Zephyr build system uses this user cache to store Zephyr compiler check
+# results which significantly improve toolchain testing performance.
+# See https://github.com/zephyrproject-rtos/zephyr/pull/7102 for details.
+#
+# Outcome:
+# The following variables will be defined when this CMake module completes:
+#
+# - USER_CACHE_DIR: User cache directory in use.
+#
+# CMake module dependencies:
+# - extensions CMake module.
+
+# Populate USER_CACHE_DIR with a directory that user applications may
+# write cache files to.
+if(NOT DEFINED USER_CACHE_DIR)
+  find_appropriate_cache_directory(USER_CACHE_DIR)
+endif()
+message(STATUS "Cache files will be written to: ${USER_CACHE_DIR}")
```

Comment:
> IMHO it is overkill to have a module for this.

I kinda felt the same when creating it, but the reason comes below.

> Let's modify the lone user (zephyr_check_compiler_flag) to do ...

Except that you will then do a `if(NOT DEFINED ...)` on each invocation on something that should just be carried out once during first stage of the CMake configure stage.

Also, this is the `USER_CACHE_DIR`, whereas the `zephyr_check_compiler_flag()` uses the `ZEPHYR_TOOLCHAIN_CAPABILITY_CACHE_DIR` which relies on the `USER_CACHE_DIR`.
In Zephyr we only use the `USER_CACHE_DIR` as base for `ZEPHYR_TOOLCHAIN_CAPABILITY_CACHE_DIR`, but in general, the `USER_CACHE_DIR` could be used for other things as well which we mention here:
https://github.com/zephyrproject-rtos/zephyr/blob/7f5eb921dc8bb1b7014886986bbffdf111b68b15/cmake/app/boilerplate.cmake#L509-L510

I think having the user cache dir available makes sense, but initializing that based on the fact that `zephyr_check_compiler_flag` has been invoked seems wrong.

Custom toolchains may not call `zephyr_check_compiler_flag` at all in which case the `USER_CACHE_DIR` would not have been set, but someone could be wanting to use `USER_CACHE_DIR` for application specific purposes.

Thus the reason having this as independent module.
I found no alternative good location.

Of course it would be possible to specify the `USER_CACHE_DIR` unconditionally as part of loading `extensions.cmake` but that would break the design of that module and be even worse than a small independent `user_cache.cmake` file.

Note, I do know out CMake build system is too poorly documented, something high on my list of things I want to improve.

Reply:
The principle that we are trying to follow with this new organization is that things should include their dependencies. That is how we get modularity.

But in this PR neither extensions.cmake nor zephyr_check_compiler_flag is including their dependency on USER_CACHE_DIR.

After making things more modular, we are at increased risk that this dependency goes unmet. Before the dependency was enforced to be met as it was only supported to execute the scripts in a single order. The one encoded in boilerplate.cmake.

We can't make things more modular, without also enforcing dependencies.

IMHO, the best tool to enforce this dependency is an ifdef on the USER_CACHE_DIR variable for each user of the variable. (Because there is only one user and therefore this is cheap).

If you insist on using modules to enforce this particular dependency then you could move the toolchain testing into a module and then have that module depend on user_cache. But IMHO it is cleaner to just do an ifdef.

But right now this dependency is un-enforced which is not OK IMHO.

Reply:
> But in this PR neither extensions.cmake nor zephyr_check_compiler_flag is including their dependency on USER_CACHE_DIR.

Fair point.
For now, `extensions.cmake` should definitely have a `include(user_cache)`.
This will also ensure that the user cache dir is established when loading extensions.

Will update user_cache and extensions.cmake so that this 

> We can't make things more modular, without also enforcing dependencies.

Agree. as long as it is a real dependency.
But not all usage are dependencies.

For example `zephyr_module` and `west` are not real dependencies of any other module, cause all other modules shall be able to work without those two modules.

> But right now this dependency is un-enforced which is not OK IMHO.

I agree.

Reply:
> Will update user_cache and extensions.cmake so that this is clear

Fixed.
`extensions.cmake` is now depending upon `user_cache` and loads this module if needed.